File: examples/5/duplicate_first_names.py
#!/usr/local/bin/python3.5 -u
# Written by andrewt@cse.unsw.edu.au for COMP2041
# Report cases where there are multiple people
# of the same same first name enrolled in a course

import fileinput, re

cfn = {}
for line in fileinput.input():
    fields = line.split('|')
    course = fields[0]
    full_name = fields[2]
    m = re.match(r'.*,\s+(\S+)', full_name)
    if not m:
        continue
    first_name = m.group(1)
    if course not in cfn:
        cfn[course] = {}
    if first_name in cfn[course]:
        cfn[course][first_name] += 1
    else:
        cfn[course][first_name] = 1

for course in sorted(cfn.keys()):
    for first_name in sorted(cfn[course].keys()):
        n = cfn[course][first_name]
        if n > 1:
            print("In %s there are %d people with the first name %s"%(course, n, first_name))

File: examples/5/duplicate_first_names.pl
#!/usr/bin/perl -w
# Written by andrewt@cse.unsw.edu.au for COMP2041
# run as duplicate_first_names.pl /home/cs2041/public_html/11s2/lec/perl/examples/enrollments
# Report cases where there are multiple people
# of the same same first name enrolled in a course

while ($line = <>) {
    @fields = split /\|/, $line;
    $course = $fields[0];
    $full_name = $fields[2];
    $full_name =~ /.*,\s+(\S+)/ or next;
    $first_name = $1;
    $cfn{$course}{$first_name}++;
}

foreach $course (sort keys %cfn) {
    foreach $first_name (sort keys %{$cfn{$course}}) {
        next if $cfn{$course}{$first_name} < 2;
        printf "In $course there are %d people with the first name $first_name\n", $cfn{$course}{$first_name};
    }
}
